### PR TITLE
feat: Add copyWith method to all models

### PR DESCRIPTION
**Pull request by M123-dev**
### What
Adds copyWith methods to all JsonSerializable models we have

My first try was to use [freezed](https://pub.dev/packages/freezed) as this can add a copyWith method to JsonSerializable but because we have a lot of customization around languages and so on I decided to go for this simpler method.

For most classes it was just adding a `@CopyWith()` decorator others needed a change in the constructor to use named parameter.

I added it to all classes in a first run, if there are some were we don't need it I'am happy to remove them

### For
https://github.com/openfoodfacts/smooth-app/pull/3059
https://github.com/openfoodfacts/smooth-app/issues/3263
...

### CODE REVIEW

File: lib/model/TaxonomyAdditive.dart [hunk 1]
```diff
@@ -62,9 +63,10 @@ enum TaxonomyAdditiveField implements OffTagged {
 ///
 /// See [OpenFoodAPIClient.getTaxonomy] for more details on how to retrieve one
 /// of these.
+@CopyWith()
 @JsonSerializable()
 class TaxonomyAdditive extends JsonObject {
-  TaxonomyAdditive(
+  TaxonomyAdditive({
```
Comment (monsieurtanuki) [inline lib/model/TaxonomyAdditive.dart:69]: Breaking change.

File: lib/model/TaxonomyAllergen.dart [hunk 2]
```diff
@@ -31,13 +32,14 @@ enum TaxonomyAllergenField implements OffTagged {
 ///
 /// See [OpenFoodAPIClient.getTaxonomy] for more details on how to retrieve one
 /// of these.
+@CopyWith()
 @JsonSerializable()
 class TaxonomyAllergen extends JsonObject {
-  TaxonomyAllergen(
+  TaxonomyAllergen({
```
Comment (monsieurtanuki) [inline lib/model/TaxonomyAllergen.dart:38]: Breaking change.

File: lib/model/TaxonomyCategory.dart [hunk 3]
```diff
@@ -58,9 +59,10 @@ enum TaxonomyCategoryField implements OffTagged {
 ///
 /// See [OpenFoodAPIClient.getTaxonomy] for more details on how to retrieve one
 /// of these.
+@CopyWith()
 @JsonSerializable()
 class TaxonomyCategory extends JsonObject {
-  TaxonomyCategory(
+  TaxonomyCategory({
```
Comment (monsieurtanuki) [inline lib/model/TaxonomyCategory.dart:65]: Breaking change.

File: lib/model/TaxonomyCountry.dart [hunk 4]
```diff
@@ -33,9 +34,18 @@ enum TaxonomyCountryField implements OffTagged {
 ///
 /// See [OpenFoodAPIClient.getTaxonomy] for more details on how to retrieve one
 /// of these.
+@CopyWith()
 @JsonSerializable()
 class TaxonomyCountry extends JsonObject {
-  TaxonomyCountry();
+  TaxonomyCountry({
```
Comment (monsieurtanuki) [inline lib/model/TaxonomyCountry.dart:40]: Did `copy_with` force you to do that?

File: lib/model/TaxonomyIngredient.dart [hunk 5]
```diff
@@ -78,9 +79,10 @@ enum TaxonomyIngredientField implements OffTagged {
 ///
 /// See [OpenFoodAPIClient.getTaxonomy] for more details on how to retrieve one
 /// of these.
+@CopyWith()
 @JsonSerializable()
 class TaxonomyIngredient extends JsonObject {
-  TaxonomyIngredient(
+  TaxonomyIngredient({
```
Comment (monsieurtanuki) [inline lib/model/TaxonomyIngredient.dart:85]: Breaking change.

File: lib/model/TaxonomyLabel.dart [hunk 6]
```diff
@@ -52,9 +53,10 @@ enum TaxonomyLabelField implements OffTagged {
 ///
 /// See [OpenFoodAPIClient.getTaxonomy] for more details on how to retrieve one
 /// of these.
+@CopyWith()
 @JsonSerializable()
 class TaxonomyLabel extends JsonObject {
-  TaxonomyLabel(
+  TaxonomyLabel({
```
Comment (monsieurtanuki) [inline lib/model/TaxonomyLabel.dart:59]: Breaking change.

File: lib/model/TaxonomyLanguage.dart [hunk 7]
```diff
@@ -32,14 +33,15 @@ enum TaxonomyLanguageField implements OffTagged {
 ///
 /// See [OpenFoodAPIClient.getTaxonomy] for more details on how to retrieve one
 /// of these.
+@CopyWith()
 @JsonSerializable()
 class TaxonomyLanguage extends JsonObject {
-  TaxonomyLanguage(
+  TaxonomyLanguage({
```
Comment (monsieurtanuki) [inline lib/model/TaxonomyLanguage.dart:39]: Breaking change.

File: lib/model/TaxonomyOrigin.dart [hunk 8]
```diff
@@ -26,9 +27,15 @@ enum TaxonomyOriginField implements OffTagged {
 ///
 /// See [OpenFoodAPIClient.getTaxonomy] for more details on how to retrieve one
 /// of these.
+@CopyWith()
 @JsonSerializable()
 class TaxonomyOrigin extends JsonObject {
-  TaxonomyOrigin();
+  TaxonomyOrigin({
```
Comment (monsieurtanuki) [inline lib/model/TaxonomyOrigin.dart:33]: Did `copy_with` force you to do that?

File: lib/model/TaxonomyPackaging.dart [hunk 9]
```diff
@@ -41,9 +42,16 @@ enum TaxonomyPackagingField implements OffTagged {
 ///
 /// See [OpenFoodAPIClient.getTaxonomy] for more details on how to retrieve one
 /// of these.
+@CopyWith()
 @JsonSerializable()
 class TaxonomyPackaging extends JsonObject {
-  TaxonomyPackaging();
+  TaxonomyPackaging({
```
Comment (monsieurtanuki) [inline lib/model/TaxonomyPackaging.dart:48]: idem

File: lib/model/TaxonomyPackagingMaterial.dart [hunk 10]
```diff
@@ -26,9 +27,15 @@ enum TaxonomyPackagingMaterialField implements OffTagged {
 ///
 /// See [OpenFoodAPIClient.getTaxonomy] for more details on how to retrieve one
 /// of these.
+@CopyWith()
 @JsonSerializable()
 class TaxonomyPackagingMaterial extends JsonObject {
-  TaxonomyPackagingMaterial();
+  TaxonomyPackagingMaterial({
```
Comment (monsieurtanuki) [inline lib/model/TaxonomyPackagingMaterial.dart:33]: idem

File: lib/model/TaxonomyPackagingRecycling.dart [hunk 11]
```diff
@@ -28,9 +29,17 @@ enum TaxonomyPackagingRecyclingField implements OffTagged {
 ///
 /// See [OpenFoodAPIClient.getTaxonomy] for more details on how to retrieve one
 /// of these.
+@CopyWith()
 @JsonSerializable()
 class TaxonomyPackagingRecycling extends JsonObject {
-  TaxonomyPackagingRecycling();
+  TaxonomyPackagingRecycling({
```
Comment (monsieurtanuki) [inline lib/model/TaxonomyPackagingRecycling.dart:35]: idem

File: lib/model/TaxonomyPackagingShape.dart [hunk 12]
```diff
@@ -26,9 +27,15 @@ enum TaxonomyPackagingShapeField implements OffTagged {
 ///
 /// See [OpenFoodAPIClient.getTaxonomy] for more details on how to retrieve one
 /// of these.
+@CopyWith()
 @JsonSerializable()
 class TaxonomyPackagingShape extends JsonObject {
-  TaxonomyPackagingShape();
+  TaxonomyPackagingShape({
```
Comment (monsieurtanuki) [inline lib/model/TaxonomyPackagingShape.dart:33]: idem

**Review by monsieurtanuki — COMMENTED** [review]
Hi @M123-dev!
A lot to review, but each time it's rather quick!

TBH, I cannot picture exactly the added value of that "CopyWith" feature, more specifically in the context of https://github.com/openfoodfacts/smooth-app/pull/3059 and https://github.com/openfoodfacts/smooth-app/issues/3263.

In those Smoothie issues, we need to "merge" two `Product`s (the second on top on the first, only for the second's non null fields), like `inputProduct.merge(product: changes)`, or `Product.merge(product1, product2)`.
And from what I saw in the generated code it's just about individual fields. Like `product1.copyWith(lang: 'de')`.
If so, that would mean that we would need to list all the fields of the second product and then use `copyWith` just on it, which is more a pain in the neck than a good feature.

And in dart, you can already write easily `Product(barcode: '1234')..lang='de'..website='https://...'`.

Therefore, I think we need at least a relevant example of the use of `copyWith`. Would you please share one?

Beyond that, you introduced slight breaking changes, but as it's for classes that are not very interesting I'm sure you'd agree to rollback your changes there, as you suggested in the OP.

**Comment by M123-dev** [comment]
Ohh, my bad @monsieurtanuki it makes sense that these methods do not make sense (at least for us), I was a bit hasty with the implementation.

**Comment by monsieurtanuki** [comment]
Thank you anyway for your attempt @M123-dev!

**Comment by M123-dev** [comment]
So looking at the problem how we actually have it. 
- We need a way to combine two objects
- Without specifying every property
- Preferably without having to write all properties
- My idea for this is to convert both passed objects to JSON and then into to a map and then just calling 
Map.addAll(), this doesn't work for nested properties though. Do you have another idea?

**Comment by monsieurtanuki** [comment]
@M123-dev Actually I think I had a specific need like that in https://github.com/openfoodfacts/smooth-app/pull/3339 (I created a `TODO` about it): a clean copy of a `Product`(like you suggested, a sort of `fromJson(toJson))`.

Should we generalize that? I don't know. For the moment we only have this specific need for `Product`.

Another grander idea would be to code a bit like in `Nutriments`: just keep the json map and create getter/setter for all "json fields".
But it would be a bit out of proportions: refactoring all our classes because for one single need in Smoothie we may find interesting to use a `copyWith`/`merge` method on some fields of `Product`...

If you feel like refactoring in general, I would suggest #450. Would impact _all_ files, with breaking changes. But the code would be cleaner afterwards.

**Pull request closed without merging**